sky/provision/kubernetes/network.py

```diff
@@ -57,39 +57,46 @@ def _open_ports_using_ingress(
     ports: List[int],
     provider_config: Dict[str, Any],
 ) -> None:
+    # Check if an ingress controller exists
     if not network_utils.ingress_controller_exists():
         raise Exception(
             'Ingress controller not found. '
             'Install Nginx ingress controller first: '
             'https://github.com/kubernetes/ingress-nginx/blob/main/docs/deploy/index.md.'  # pylint: disable=line-too-long
         )
 
-    for port in ports:
-        service_name = f'{cluster_name_on_cloud}-skypilot-service--{port}'
-        ingress_name = f'{cluster_name_on_cloud}-skypilot-ingress--{port}'
-        path_prefix = _PATH_PREFIX.format(
-            cluster_name_on_cloud=cluster_name_on_cloud, port=port)
+    # Prepare service names, ports,  for template rendering
+    service_details = [
+        (f'{cluster_name_on_cloud}-skypilot-service--{port}', port,
+         _PATH_PREFIX.format(cluster_name_on_cloud=cluster_name_on_cloud,
+                             port=port).rstrip('/').lstrip('/'))
+        for port in ports
+    ]
 
-        content = network_utils.fill_ingress_template(
-            namespace=provider_config.get('namespace', 'default'),
-            path_prefix=path_prefix,
-            service_name=service_name,
-            service_port=port,
-            ingress_name=ingress_name,
-            selector_key='skypilot-cluster',
-            selector_value=cluster_name_on_cloud,
-        )
+    # Generate ingress and services specs
+    content = network_utils.fill_ingress_template(
+        namespace=provider_config.get('namespace', 'default'),
+        service_details=service_details,
+        ingress_name=f'{cluster_name_on_cloud}-skypilot-ingress',
+        selector_key='skypilot-cluster',
+        selector_value=cluster_name_on_cloud,
+    )
+
+    # Create or update services based on the generated specs
+    for service_name, service_spec in content['services_spec'].items():
         network_utils.create_or_replace_namespaced_service(
             namespace=provider_config.get('namespace', 'default'),
             service_name=service_name,
-            service_spec=content['service_spec'],
-        )
-        network_utils.create_or_replace_namespaced_ingress(
-            namespace=provider_config.get('namespace', 'default'),
-            ingress_name=ingress_name,
-            ingress_spec=content['ingress_spec'],
+            service_spec=service_spec,
         )
 
+    # Create or update the single ingress for all services
+    network_utils.create_or_replace_namespaced_ingress(
+        namespace=provider_config.get('namespace', 'default'),
+        ingress_name=f'{cluster_name_on_cloud}-skypilot-ingress',
+        ingress_spec=content['ingress_spec'],
+    )
```

Comment:
Just curious, if we are changing the ingress for all the ports in batch, will this cause the other ports being unavailable for a while when a new port is added?

Also, just to confirm, when adding a new port, the original endpoint for the existing ports do not change?

Reply:
>Just curious, if we are changing the ingress for all the ports in batch, will this cause the other ports being unavailable for a while when a new port is added?

Yes, that's true. Nginx will do a hot reload whenever a new ingress rule is created. As a result, ports may be unavailable for a short duration before getting restotred.

> Also, just to confirm, when adding a new port, the original endpoint for the existing ports do not change?

Yes, the original endpoint will remain the same, since the k8s service remains the same.